api/envoy/service/mutation/v3alpha/mutation.proto

```diff
@@ -0,0 +1,302 @@
+// to a mutation server.
+// [#next-free-field: 8]
+message MutationRequest {
+  // If true, then Envoy requires that the server send one MutationResponse message
+  // back up the stream in response to this message -- otherwise, Envoy will
+  // never resume processing the filter chain. In practice, this will always
+  // be "true" unless the server or the user has configured the filter
+  // to run in "async mode."
+  bool response_required = 1;
+
+  // Information about the HTTP request, sent when the decodeHeaders callback
+  // is invoked on the Envoy filter.
+  Request request = 2;
+
+  // Information about the HTTP response, sent when the encodeHeaders callback
+  // is invoked on the Envoy filter.
+  Response response = 3;
+
+  // A chunk of the HTTP request body. This is only sent if the server
+  // requested it.
+  Body request_body = 4;
+
+  // A chunk of the HTTP response body, if the server requested it.
+  Body response_body = 5;
+
+  // The HTTP trailers for the request path. These are only sent if
+  // the server requested them.
+  config.core.v3.HeaderMap request_trailers = 6;
+
+  // The HTTP trailers for the response path, sent if the server
+  // requested them.
+  config.core.v3.HeaderMap response_trailers = 7;
+}
+
+// The following are commands that the external server can send back to
+// the proxy to direct the next phase of processing.
+
+// The server is done with the request and response, and no further
+// filters should be invoked. This is typically used with the
+// "CompleteResponse" to return an error.
+message StopCommand {
+}
+
+// The server is done with the request. Continue processing
+// and do not consult the server again for this request.
+message ContinueRequestCommand {
+}
+
+// The server is done with the response. Do not consult the server
+// again.
+message ContinueResponseCommand {
+}
+
+// Request that Envoy next send a Body message downstream
+// containing either the next chunk of the request body available, or an end-of-stream.
+// Once this is returned, it will no longer be possible to modify the
+// message headers.
+message SendRequestBodyChunkCommand {
+}
+
+// Send the next chunk of the response body.
+message SendResponseBodyChunkCommand {
+}
+
+// Request that Envoy will buffer the entire request message body in memory
+// subject to the currently-configured body buffer limit, and send a
+// single Body message downstream containing the entire message body.
+// It is still possible for the server to modify the message headers
+// with the next upstream message.
+message SendBufferedRequestBodyCommand {
+}
+
+// Request that Envoy buffer and send the response body in a single chunk.
+message SendBufferedResponseBodyCommand {
+}
+
+// Request that Envoy will send the HTTP request trailers when they are
+// available. Once trailers have been requested, no more bodies
+// will be delivered, so this must be sent after all desired body chunks are
+// desired.
```

Comment:
How are no trailers / end of stream indicated?

Reply:
I'll update the comment to clarify that an empty set of trailers will be sent if trailers are requested but none are present. (And I'll figure out later whether the HTTP filter model is actually capable of doing that.)